### PR TITLE
add groupBy

### PR DESCRIPTION
**Pull request by danking**
Also scalatest has the worst matching infrastructure, ever.

cc @lfrancioli @konradjk

### CODE REVIEW

File: src/test/scala/is/hail/methods/ExprSuite.scala [hunk 1]
```diff
@@ -585,6 +588,56 @@ class ExprSuite extends SparkSuite {
     // FIXME: parser should accept minimum Long/Int literals
     // assert(eval[Long](Long.MinValue.toString+"L").contains(Long.MinValue))
     // assert(eval[Long](Long.MinValue.toString+"l").contains(Long.MinValue))
+
+    {
+      val x = eval[Map[String,IndexedSeq[Int]]]("[1,2,3,4,5].groupBy(k => if (k % 2 == 0) \"even\" else \"odd\")")
+      assert(x.isDefined)
+      x.get.keySet should contain theSameElementsAs Seq("even", "odd")
+      x.get("even") should contain theSameElementsAs Seq(2,4)
+      x.get("odd") should contain theSameElementsAs Seq(1,3,5)
```
Comment (danking) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: I want to actually write this as:

```scala
(eval[Map[String,IndexedSeq[Int]]]("[1,2,3,4,5].groupBy(k => if (k % 2 == 0) \"even\" else \"odd\")")
  should contain (contain theSameElementsAs ("even" -> (contain theSameElementsAs Seq(2,4)),
                                             "odd" -> (contain theSameElementsAs Seq(1,3,5)))))
```

or really:

```scala
(eval[Map[String,IndexedSeq[Int]]]("[1,2,3,4,5].groupBy(k => if (k % 2 == 0) \"even\" else \"odd\")")
  shouldBe Some(Map("even" -> exactlyInAnyOrder(2,4), "odd" -> exactlyInAnyOrder(1,3,5))))
```

But scala test matchers are not compassable at all.

File: docs/reference/HailExpressionLanguage.md [hunk 2]
```diff
@@ -83,6 +83,7 @@ Several Hail commands provide the ability to perform a broad array of computatio
      - forall: `arr.forall(v => expr)` -- Returns a boolean which is true if the array is empty, or `expr` evaluates to `true` for **every** element
      - sort: `arr.sort([ascending])` -- Returns a new array with the same elements in ascending order according to their value, which must be numeric or string. For descending order, use `arr.sort(false)`. Missing elements are always placed at the end.
      - sortBy: `arr.sortBy(v => expr[,ascending])` -- Returns a new array with the same elements in ascending order according to the value of `expr`, which must be numeric or string. For descending order, use `arr.sortBy(v => expr, false)`. Elements with missing `expr` values are always placed at the end.
+     - groupBy: `arr.groupBy(v => v.k)` -- Returns a Dict mapping from keys specified by the lambda argument to an Array of the values associated with that key. The keys must be Strings.
```
Comment (cseed) [inline docs/reference/HailExpressionLanguage.md]: You should rebase this off Tim's key-parameterized `Dict`s.

Comment (cseed) [inline docs/reference/HailExpressionLanguage.md]: So here we have `Array[T].groupBy[U](T => U): Dict[U, Array[T]]` and also for `Set`.  I think there is hope for group by on `Aggregable`s.  Tricky.  One option would be `Aggregable[T].groupBy(T => U, Aggregable[T] => W): Dict[U, W]`.

File: src/test/scala/is/hail/methods/ExprSuite.scala (hunk 1 above)
Comment (cseed) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: > But scala test matchers are not compassable at all.

What?  Really?  Trying to embed English grammar in code doesn't compose?  I'm so surprised.

File: src/test/scala/is/hail/methods/ExprSuite.scala [hunk 3]
```diff
@@ -585,6 +588,56 @@ class ExprSuite extends SparkSuite {
     // FIXME: parser should accept minimum Long/Int literals
     // assert(eval[Long](Long.MinValue.toString+"L").contains(Long.MinValue))
     // assert(eval[Long](Long.MinValue.toString+"l").contains(Long.MinValue))
+
+    {
+      val x = eval[Map[String,IndexedSeq[Int]]]("[1,2,3,4,5].groupBy(k => if (k % 2 == 0) \"even\" else \"odd\")")
+      assert(x.isDefined)
+      x.get.keySet should contain theSameElementsAs Seq("even", "odd")
```
Comment (cseed) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: I won't fight this, but I have to say I this this whole let's-write-tests-as-if-English things is horse shit.  Let's give it a fancy name some people will confuse it with a serious development methodology!  Yeah!  This is me not fighting.

I do admit our testing error messages are shitty, and if this fixes that, it is an improvement.

File: docs/reference/HailExpressionLanguage.md (hunk 2 above)
Comment (danking) [inline docs/reference/HailExpressionLanguage.md]: The Java 8 Streams library [adopts the same interface](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#groupingBy-java.util.function.Function-java.util.stream.Collector-). I agree, it's the right way to do `Aggregable.groupBy`.

File: src/test/scala/is/hail/methods/ExprSuite.scala (hunk 3 above)
Comment (danking) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: I agree whole-hearted, writing tests (or any computer algorithm) as natural English is shit. I just want intelligent matchers that compose and produce useful error messages. I'm disappointed with the response of the scalatest developer to https://github.com/scalatest/scalatest/issues/471, which AFAICT, reaffirms the commitment to english-language tests, not to tests that explain what went wrong.

I suspect we could build a matcher library to do what I want below. Hopefully that would also compose well with the quick check library.

Comment (danking) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: It's also worth noting that what drove me into this painful corner of testing libraries was that I realized that I didn't really want to require any ordering on the value arrays, especially in the set case.

**Comment by danking** [comment]
Don't merge until #1367 lands

**Pull request merged**